packages/cli/src/routeGeneration/templates/express.hbs

```diff
@@ -104,51 +106,63 @@ export function RegisterRoutes(app: express.Router) {
-                    error.status = error.status || 401;
-                    next(error)
-                }
-            }
+    function authenticateMiddleware(security: TsoaRoute.Security[] = []) {
+        return async function runAuthenticationMiddleware(request: any, _response: any, next: any) {
 
             // WARNING: This file was auto-generated with tsoa. Please do not modify it. Re-run tsoa to re-generate this file: https://github.com/lukeautry/tsoa
 
+            // keep track of failed auth attempts so we can hand back the most
+            // recent one.  This behavior was previously existing so preserving it
+            // here
+            const failedAttempts: any[] = [];
+            const pushAndRethrow = (error: any) => {
+                failedAttempts.push(error);
+                throw error;
+            };
+
+            const secMethodOrPromises: Promise<any>[] = [];
             for (const secMethod of security) {
                 if (Object.keys(secMethod).length > 1) {
-                    let promises: Promise<any>[] = [];
+                    const secMethodAndPromises: Promise<any>[] = [];
 
                     for (const name in secMethod) {
-                        promises.push(expressAuthentication(request, name, secMethod[name]));
+                        secMethodAndPromises.push(
+                            expressAuthentication(request, name, secMethod[name])
+                                .catch(pushAndRethrow)
+                        );
                     }
 
-                    Promise.all(promises)
-                        .then((users) => { succeed(users[0]); })
-                        .catch(fail);
+                    // WARNING: This file was auto-generated with tsoa. Please do not modify it. Re-run tsoa to re-generate this file: https://github.com/lukeautry/tsoa
+
+                    secMethodOrPromises.push(Promise.all(secMethodAndPromises)
+                        .then(users => { return users[0]; }));
                 } else {
                     for (const name in secMethod) {
-                        expressAuthentication(request, name, secMethod[name])
-                            .then(succeed)
-                            .catch(fail);
+                        secMethodOrPromises.push(
+                            expressAuthentication(request, name, secMethod[name])
+                                .catch(pushAndRethrow)
+                        );
                     }
                 }
             }
+
+            // WARNING: This file was auto-generated with tsoa. Please do not modify it. Re-run tsoa to re-generate this file: https://github.com/lukeautry/tsoa
+
+            try {
+                request['user'] = await promiseAny(secMethodOrPromises);
+                next();
+            }
+            catch(err) {
+                // Show most recent error as response
+                const error = failedAttempts.pop();
```

Comment:
Instead of manual tracking, err.errors.pop()

Reply:
<details>
<summary>Here's the combined diff</summary>

```diff
diff --git a/packages/cli/src/routeGeneration/templates/express.hbs b/packages/cli/src/routeGeneration/templates/express.hbs
index 222b28b..1ae0354 100644
--- a/packages/cli/src/routeGeneration/templates/express.hbs
+++ b/packages/cli/src/routeGeneration/templates/express.hbs
@@ -9,7 +9,7 @@ import { {{name}} } from '{{modulePath}}';
 {{#if authenticationModule}}
 import { expressAuthentication } from '{{authenticationModule}}';
 // @ts-ignore - no great way to install types from subpackage
-const promiseAny = require('promise.any');
+import * as promiseAny from 'promise.any';
 {{/if}}
 {{#if iocModule}}
 import { iocContainer } from '{{iocModule}}';
@@ -114,15 +114,6 @@ export function RegisterRoutes(app: express.Router) {
 
             // WARNING: This file was auto-generated with tsoa. Please do not modify it. Re-run tsoa to re-generate this file: https://github.com/lukeautry/tsoa
 
-            // keep track of failed auth attempts so we can hand back the most
-            // recent one.  This behavior was previously existing so preserving it
-            // here
-            const failedAttempts: any[] = [];
-            const pushAndRethrow = (error: any) => {
-                failedAttempts.push(error);
-                throw error;
-            };
-
             const secMethodOrPromises: Promise<any>[] = [];
             for (const secMethod of security) {
                 if (Object.keys(secMethod).length > 1) {
@@ -131,7 +122,6 @@ export function RegisterRoutes(app: express.Router) {
                     for (const name in secMethod) {
                         secMethodAndPromises.push(
                             expressAuthentication(request, name, secMethod[name])
-                                .catch(pushAndRethrow)
                         );
                     }
 
@@ -143,7 +133,6 @@ export function RegisterRoutes(app: express.Router) {
                     for (const name in secMethod) {
                         secMethodOrPromises.push(
                             expressAuthentication(request, name, secMethod[name])
-                                .catch(pushAndRethrow)
                         );
                     }
                 }
@@ -157,7 +146,7 @@ export function RegisterRoutes(app: express.Router) {
             }
             catch(err) {
                 // Show most recent error as response
-                const error = failedAttempts.pop();
+                const error = err.errors.pop();
                 error.status = error.status || 401;
                 next(error);
             }
diff --git a/packages/cli/src/routeGeneration/templates/hapi.hbs b/packages/cli/src/routeGeneration/templates/hapi.hbs
index ddb3251..c1822b6 100644
--- a/packages/cli/src/routeGeneration/templates/hapi.hbs
+++ b/packages/cli/src/routeGeneration/templates/hapi.hbs
@@ -9,7 +9,7 @@ import { {{name}} } from '{{modulePath}}';
 {{#if authenticationModule}}
 import { hapiAuthentication } from '{{authenticationModule}}';
 // @ts-ignore - no great way to install types from subpackage
-const promiseAny = require('promise.any');
+import * as promiseAny from 'promise.any';
 {{/if}}
 {{#if iocModule}}
 import { iocContainer } from '{{iocModule}}';
@@ -145,15 +145,6 @@ export function RegisterRoutes(server: any) {
 
             // WARNING: This file was auto-generated with tsoa. Please do not modify it. Re-run tsoa to re-generate this file: https://github.com/lukeautry/tsoa
 
-            // keep track of failed auth attempts so we can hand back the most
-            // recent one.  This behavior was previously existing so preserving it
-            // here
-            const failedAttempts: any[] = [];
-            const pushAndRethrow = (error: any) => {
-                failedAttempts.push(error);
-                throw error;
-            };
-
             const secMethodOrPromises: Promise<any>[] = [];
             for (const secMethod of security) {
                 if (Object.keys(secMethod).length > 1) {
@@ -162,7 +153,6 @@ export function RegisterRoutes(server: any) {
                     for (const name in secMethod) {
                         secMethodAndPromises.push(
                             hapiAuthentication(request, name, secMethod[name])
-                                .catch(pushAndRethrow)
                         );
                     }
 
@@ -174,7 +164,6 @@ export function RegisterRoutes(server: any) {
                     for (const name in secMethod) {
                         secMethodOrPromises.push(
                             hapiAuthentication(request, name, secMethod[name])
-                                .catch(pushAndRethrow)
                         );
                     }
                 }
@@ -188,7 +177,7 @@ export function RegisterRoutes(server: any) {
             }
             catch(err) {
                 // Show most recent error as response
-                const error = failedAttempts.pop();
+                const error = err.errors.pop();
                 if (isBoom(error)) {
                     throw error;
                 }
diff --git a/packages/cli/src/routeGeneration/templates/koa.hbs b/packages/cli/src/routeGeneration/templates/koa.hbs
index 5147456..4855634 100644
--- a/packages/cli/src/routeGeneration/templates/koa.hbs
+++ b/packages/cli/src/routeGeneration/templates/koa.hbs
@@ -13,7 +13,7 @@ import { {{name}} } from '{{modulePath}}';
 {{#if authenticationModule}}
 import { koaAuthentication } from '{{authenticationModule}}';
 // @ts-ignore - no great way to install types from subpackage
-const promiseAny = require('promise.any');
+import * as promiseAny from 'promise.any';
 {{/if}}
 {{#if iocModule}}
 import { iocContainer } from '{{iocModule}}';
@@ -114,15 +114,6 @@ export function RegisterRoutes(router: KoaRouter) {
 
             // WARNING: This file was auto-generated with tsoa. Please do not modify it. Re-run tsoa to re-generate this file: https://github.com/lukeautry/tsoa
 
-            // keep track of failed auth attempts so we can hand back the most
-            // recent one.  This behavior was previously existing so preserving it
-            // here
-            const failedAttempts: any[] = [];
-            const pushAndRethrow = (error: any) => {
-                failedAttempts.push(error);
-                throw error;
-            };
-
             const secMethodOrPromises: Promise<any>[] = [];
             for (const secMethod of security) {
                 if (Object.keys(secMethod).length > 1) {
@@ -131,7 +122,6 @@ export function RegisterRoutes(router: KoaRouter) {
                     for (const name in secMethod) {
                         secMethodAndPromises.push(
                             koaAuthentication(context.request, name, secMethod[name])
-                                .catch(pushAndRethrow)
                         );
                     }
 
@@ -141,7 +131,6 @@ export function RegisterRoutes(router: KoaRouter) {
                     for (const name in secMethod) {
                         secMethodOrPromises.push(
                             koaAuthentication(context.request, name, secMethod[name])
-                                .catch(pushAndRethrow)
                         );
                     }
                 }
@@ -157,7 +146,7 @@ export function RegisterRoutes(router: KoaRouter) {
             }
             catch(err) {
                 // Show most recent error as response
-                const error = failedAttempts.pop();
+                const error = err.errors.pop();
                 context.status = error.status || 401;
                 context.throw(context.status, error.message, error);
             }
```
</details>

Reply:
if we pop the error off of the aggregated error list, I believe this will always be the error from the last auth decorator.  The failedAttempts array tracks the rejections in the order that they failed rather than the order they were specified.  I do like the simplicity of not worrying about that but it is a breaking change and I've always thought the behavior tended to choose the "right" error to report on in most cases. I could be wrong though. Since you're getting ready to rev the major we could make that change as part of that assuming we also get something in to control the order in a timely manner.

Reply:
Seems about right, my brain actively tries to block that out as it's just not how I would want this to (have) work(ed).
I'll try to build on your PR and propagate a `TsoaAuthenticationError` that extends `AggregateError` for v4, I think this has to be the long-term plan.
However, I still can't reproduce the tests failing with `import * as`. Any hints there?